c++/src/kj/compat/tls-test.c++

```diff
@@ -818,6 +822,241 @@ KJ_TEST("TLS client certificate verification") {
+    return acceptImpl();
+  }
+
+  Promise<AuthenticatedStream> acceptAuthenticated() override {
+    return acceptImpl().then([](auto stream) -> AuthenticatedStream {
+      return { kj::mv(stream), LocalPeerIdentity::newInstance({}) };
+    });
+  }
+
+  uint getPort() override {
+    return 0;
+  }
+  void getsockopt(int, int, void*, uint*) override {}
+  void setsockopt(int, int, const void*, uint) override {}
+
+  Promise<Own<AsyncIoStream>> connect() {
+    // Mock a new successful connection to our receiver.
+    return connectImpl();
+  }
+
+  Promise<void> badConnect() {
+    // Mock a new failed connection to our receiver.
+    return connectImpl(KJ_EXCEPTION(DISCONNECTED, "Pipes are leaky")).ignoreResult();
+  }
+
+private:
+  Promise<Own<AsyncIoStream>> acceptImpl() {
+    if (clientRequests.empty()) {
+      KJ_ASSERT(!serverFulfiller);
+      auto paf = newPromiseAndFulfiller<void>();
+      serverFulfiller = kj::mv(paf.fulfiller);
+      return paf.promise.then([this] {
+        return acceptImpl();
+      });
+    }
+
+    // This is accepting in FILO order, it shouldn't matter in practice.
+    auto request = kj::mv(clientRequests.back());
+    clientRequests.removeLast();
+
+    KJ_IF_MAYBE(exception, kj::mv(request.maybeException)) {
+      request.clientFulfiller = nullptr;  // The other end had an issue, break the promise.
+      return kj::mv(*exception);
+    } else {
+      auto pipe = provider.newTwoWayPipe();
+      request.clientFulfiller->fulfill(kj::mv(pipe.ends[0]));
+      return kj::mv(pipe.ends[1]);
+    }
+  }
+
+  Promise<Own<AsyncIoStream>> connectImpl(Maybe<Exception> maybeException = nullptr) {
+    auto paf = newPromiseAndFulfiller<Own<AsyncIoStream>>();
+    clientRequests.add(ClientRequest{ kj::mv(maybeException), kj::mv(paf.fulfiller) });
+
+    if (auto fulfiller = kj::mv(serverFulfiller)) {
+      fulfiller->fulfill();
+    }
+
+    return kj::mv(paf.promise);
+  }
+
+  AsyncIoProvider& provider;
+
+  Own<PromiseFulfiller<void>> serverFulfiller;
+  Vector<ClientRequest> clientRequests;
+};
+
+class TlsReceiverTest final: public TlsTest {
+  // TlsReceiverTest augments TlsTest to test TlsConnectionReceiver.
+public:
+  TlsReceiverTest(): TlsTest() {
+    auto baseReceiverPtr = kj::heap<MockConnectionReceiver>(*io.provider);
+    baseReceiver = baseReceiverPtr.get();
+    receiver = tlsServer.wrapPort(kj::mv(baseReceiverPtr));
+  }
+
+  Own<ConnectionReceiver> receiver;
+  MockConnectionReceiver* baseReceiver;
+};
+
+KJ_TEST("TLS receiver basics") {
+  TlsReceiverTest test;
+  ErrorNexus e;
+
+  test.receiver->run();
+
+  auto clientPromise = test.baseReceiver->connect()
+                           .then([&](auto stream) {
+                             return test.tlsClient.wrapClient(kj::mv(stream), "example.com");
+                           })
+                           .then([&](auto client) {
+                             return test.writeToServer(kj::mv(client));
```

Comment:
Not important, but I'm curious is there some reason why this doesn't use the `test.testConnection()` helper you added?

Reply:
It's mostly that testConnection takes a StreamPtr and not a Promise<StreamPtr>. I was hoping that distinguishing between async and sync signatures would make them easier to follow. That said, I don't want to pretend I spent that long thinking about it :).